src/webgpu/api/validation/layout_shader_compat.spec.ts

```diff
@@ -1,14 +1,318 @@
+      }
+      case 'readonlyStorageTex': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          storageTexture: {
+            format: 'r32float',
+            access: 'read-only',
+          },
+        };
+        break;
+      }
+      case 'writeonlyStorageTex': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          storageTexture: {
+            format: 'r32float',
+            access: 'write-only',
+          },
+        };
+        break;
+      }
+      case 'readwriteStorageTex': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          storageTexture: {
+            format: 'r32float',
+            access: 'read-write',
+          },
+        };
+        break;
+      }
+    }
+    return this.device.createPipelineLayout({
+      bindGroupLayouts: [
+        this.device.createBindGroupLayout({
+          entries: [bindGroupLayoutEntry],
+        }),
+      ],
+    });
+  }
+
+  GetBindableResourceShaderDeclaration(bindableResource: BindableResourceType): string {
+    switch (bindableResource) {
+      case 'compareSamp':
+        return 'var tmp : sampler_comparison';
+      case 'filtSamp':
+      case 'nonFiltSamp':
+        return 'var tmp : sampler';
+      case 'sampledTex':
+        return 'var tmp : texture_2d<f32>';
+      case 'sampledTexMS':
+        return 'var tmp : texture_multisampled_2d<f32>';
+      case 'storageBuf':
+        return 'var<storage, read_write> tmp : vec4u';
+      case 'readonlyStorageBuf':
+        return 'var<storage, read> tmp : vec4u';
+      case 'uniformBuf':
+        return 'var<uniform> tmp : vec4u;';
+      case 'readonlyStorageTex':
+        return 'var tmp : texture_storage_2d<r32float, read>';
+      case 'writeonlyStorageTex':
+        return 'var tmp : texture_storage_2d<r32float, write>';
+      case 'readwriteStorageTex':
+        return 'var tmp : texture_storage_2d<r32float, read_write>';
+    }
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('pipeline_layout_shader_exact_match')
+  .desc(
+    `
+  Test that the binding type in the pipeline layout must match the related declaration in shader.
+  Note that read-write storage textures in the pipeline layout can match write-only storage textures
+  in the shader.
+  `
+  )
+  .params(u =>
+    u
+      .combine('bindingInPipelineLayout', kBindableResources)
+      .combine('bindingInShader', kBindableResources)
+      .beginSubcases()
+      .combine('pipelineLayoutVisibility', kShaderStageCombinations)
+      .combine('shaderStageWithBinding', kShaderStages)
+      .combine('isBindingStaticallyUsed', [true, false] as const)
+      .unless(
+        p =>
+          // We don't test using non-filtering sampler in shader because it has the same declaration
+          // as filtering sampler.
+          p.bindingInShader === 'nonFiltSamp' ||
+          ((p.bindingInPipelineLayout === 'writeonlyStorageTex' ||
```

Comment:
nit: Maybe we could have a helper bindingCompatibleWithStage function to make things slightly more readable.